src/sgame/sg_admin.cpp

```diff
@@ -2350,6 +2358,93 @@ bool G_admin_setlevel( gentity_t *ent )
 	return true;
 }
 
+bool G_admin_slap( gentity_t *ent )
+{
+	int            pid;
+	double         damage = 0.0f; // in case no damage is specified
+	char           name[ MAX_NAME_LENGTH ], err[ MAX_STRING_CHARS ];
+	gentity_t      *vic;
+	vec3_t         dir;
+
+	if ( level.intermissiontime )
+	{
+		return false;
+	}
+
+	if ( trap_Argc() < 2 )
+	{
+		ADMP( QQ( N_( "^3slap:^* usage: slap [^3name|slot#|admin#^7] (^3damage^7)" ) ) ); 
+		return false; // todo: G_admin_helptext( "slap" ) instead of this copypasta
+	}
+
+	trap_Argv( 1, name, sizeof( name ) );
+
+	pid = G_ClientNumberFromString( name, err, sizeof( err ) );
+
+	if ( pid == -1 )
+	{
+		ADMP( va( "%s %s %s", QQ( "^3$1$:^* $2t$" ), "slap", Quote( err ) ) );
+		return false; // none or >1 possible matches. stop here.
+	}
+
+	vic = &g_entities[ pid ];
+
+	if ( !admin_higher( ent, vic ) )
+	{
+		ADMP( va( "%s %s", QQ( N_( "^3$1$:^* sorry, but your intended victim has a higher admin"
+		          " level than you" ) ), "slap" ) ); // todo: move this print to a single helper function
+		return false; 
+	}
+
+	const HealthComponent* health = vic->entity->Get<HealthComponent>();
+	if ( G_Team( vic ) == TEAM_NONE || !health ) // do NOT use health->Alive() here
+	{                                            // or it will crash when used on bots
+		ADMP( va( "%s %s", QQ( N_( "^3$1$:^* cannot slap spectators or dead players!" ) ), "slap" ) );
+		return false;
+	}
+
+	dir[0] = crandom();
+	dir[1] = crandom();
+	dir[2] = random();
+
+	// from G_Knockback ...
+	vec3_t kvel;
+	float mass = BG_Class( vic->client->pers.classSelection )->health;
+	// health != mass however for the purposes of this functionn
+	// using the max health of the class works a lot better
```

Comment:
Please be clear that we use health as mass to better handle the pure knockback case. That's why we're duplciating this code instead of using the knockbackcomponent.